### DISCUSSION TITLE
Database Function JOINS not returning any result

### DISCUSSION THREAD

**Discussion by calebcng** [supabase/supabase#22706] [Questions] [answered]
I am trying to create a database function that queries all posts made from a user's network of friends. My table setup is as follows:

**table: users**
- id: uuid
- handle: text
- display_name: text
- photo_url: text

**table: posts**
- id: uuid
- author_id: uuid
- content: text

**table: friendships**
- id: uuid
- sending_user_id: uuid
- receiving_user_id: uuid
- status: enumerated_type (pending, accepted, declined)

I was able to create a materialized view that achieves what I am trying to achieve using the SQL query below. 
```
create or replace view
  newsfeed_view as
select
  p.*,
  u.handle as author_handle,
  u.display_name as author_display_name,
  u.photo_url as author_photo_url,
  f.sending_user_id as sending_user_id,
  f.receiving_user_id as receiving_user_id
from
  posts p
  left join users u on p.author_id = u.id
  inner join friendships f on (f.receiving_user_id = p.author_id or f.sending_user_id = p.author_id)
```

However, when I try to do the same thing except using a database function, the resulting query returns no results (while the materialized view is showing results).
```
create or replace function get_newsfeed(user_id uuid)
returns setof newsfeed_view
language sql
as $$
  select
    p.*,
    u.handle as author_handle,
    u.display_name as author_display_name,
    u.photo_url as author_photo_url,
    f.sending_user_id as sending_user_id,
    f.receiving_user_id as receiving_user_id
  from
    posts p
    left join users u on p.author_id = u.id
    inner join friendships f on (f.receiving_user_id = p.author_id or f.sending_user_id = p.author_id)
$$;
```

When I change to using a `left join` instead of the `inner join`, it is able to return results, indicating that the inner join with the friendships table was not able to find any matches using the function as opposed to the view. 

Does anyone have any tips or pointers as to what I may be missing?

Thank you!

**Comment by GaryAustin1** [comment, marked as answer]
Not looking thru the details as late, but usually would look to meeting RLS on all tables involved IF you are calling these from the REST API.  If not ignore this post.

**Reply by calebcng** [reply]
Yes! Upon your advice, I took a look and indeed it seems to be related to the RLS policy I had on the friendships table. When I changed the `SELECT` policy to be true, I was able to get the expected matches when calling the function.
```sql
alter policy "Allow select to friendships table"
on "public"."friendships"
to public
using (
  -- Previous policy:
  ((sending_user_id = auth.uid()) OR (receiving_user_id = auth.uid()))

  -- Updated policy:
  true 
);
```

Do you have any suggestions of how I could either:
- Modify the policy on this table to still retain some level of security similar to what I had originally intended?
- Modify the function so that it can still be compatible with my original RLS policy (similar to how the materialized view was able to work)?

Thank you again!

Note: I am calling this function using the _Remote Procedure Call_ feature from my Dart code as follows:
```dart
final requests = await _client.rpc('get_newsfeed', params: {'user_id': user.id}).order('last_updated', ascending: false);
```

**Reply by GaryAustin1** [reply]
So I assume materialized view works because it bypasses RLS.  So you need to be comfortable with that.
You could change your postgres function to `security definer` type and it will also bypass RLS and the tables it uses.  But you need to be comfortable with anyone call that function and not having RLS on what it can use, or take further security measures in or on the function.
